### PR TITLE
Support closures and records in outbound live references.

### PR DESCRIPTION
**Pull request by polina-c**
RELEASE_NOTE_EXCEPTION=[behind flag]

https://github.com/flutter/devtools/issues/4962

https://user-images.githubusercontent.com/12115586/222518390-fdaa5123-e113-4d43-8a59-4877cc16f652.mov

### CODE REVIEW

File: packages/devtools_app/lib/src/shared/diagnostics/dart_object_node.dart [hunk 1]
```diff
@@ -273,23 +274,31 @@ class DartObjectNode extends TreeNode<DartObjectNode> {
 
   @override
   bool get isExpandable {
+    final theRef = ref;
```
Comment (elliette) [inline packages/devtools_app/lib/src/shared/diagnostics/dart_object_node.dart:277]: I don't think we need to assign this to a new variable

File: packages/devtools_app/lib/src/shared/diagnostics/dart_object_node.dart [hunk 2]
```diff
@@ -273,23 +274,31 @@ class DartObjectNode extends TreeNode<DartObjectNode> {
...
+    final instanceRef = theRef?.instanceRef;
+    if (theRef is ObjectReferences) {
+      if (theRef.instanceRef?.length == 0) return false;
```
Comment (elliette) [inline packages/devtools_app/lib/src/shared/diagnostics/dart_object_node.dart:280]: Could we use `instanceRef` here (instead of `theRef.instanceRef`)?

**Review by elliette — APPROVED** [review]
LGTM with a couple comments

File: packages/devtools_app/lib/src/shared/diagnostics/dart_object_node.dart (hunk 1 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/shared/diagnostics/dart_object_node.dart:277]: I need type promotion to access theRef.refNodeType little later

Comment (elliette) [reply packages/devtools_app/lib/src/shared/diagnostics/dart_object_node.dart:277]: Ah got it. Optional suggestion:

```
if (ref is ObjectReferences) {
  objectRef = ref as ObjectReferences;
  ... 
  return objectRef.refNodeType...
}

**Pull request merged**